### DISCUSSION TITLE
Parallel calling of lib's functions

### DISCUSSION THREAD

**Discussion by akrigator** [bashly-framework/bashly#505] [Q&A] [answered]
Hello, i've a small project which align TestRail API  to some kind of CLI toolkit. Recently i have discovered Bashly and have try to use it for my tiny project. One of the feature of my project is that it allows to process multiple items at once. For example, API `get case` request can process only one item at once, however in
 cli https://github.com/akrigator/testrail-cli/blob/master/cli/cases/get_case i use `parallel` for implentaion an ability to get several cases at once. But since bashly packs all lib methods to bash-functions, they became unavailable for execution in parallel calling.
Does someone use any kind of parallelism for functions which can be applied for lib's function?

**Comment by DannyBen** [comment, marked as answer]
I have no experience with `parallel`, but at least one solution seems to be possible; Define the *single* action you want to perform as a [private command](https://bashly.dannyb.co/configuration/command/#private), and simply run it through `parallel`.

**Reply by akrigator** [reply]
Yeah, such approach should work

**Comment by DannyBen** [comment]
Minimal test:

```yaml
# bashly.yml
name: cli

commands:
- name: test
- name: internal
  private: true
  args:
  - name: messages
    repeatable: true
    required: true
```

```bash
# src/test_command.sh
parallel ./cli internal ::: 1 2 3

# src/internal_command.sh
eval "messages=(${args['messages']})"
echo "===> ${messages[*]}"
```

then running:

```
$ ./cli test
===> 1
===> 2
===> 3
```

**Reply by akrigator** [reply]
Something like that i've imagine, except one thing in `parallel ./cli internal ::: 1 2 3` -  i guess that it's better to reference as `cli` instead `./cli` and add executable's path to fpath  env var. To be able run script from everywhere directory

**Reply by DannyBen** [reply]
> cli instead ./cli

of course
